### PR TITLE
feat(config): Update to import new config files

### PR DESCRIPTION
**Pull request by Wentao-Kuang**

### CODE REVIEW

File: src/provider.config.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+   */
+  async reconcile(): Promise<void> {
+    const tileSetId = Config.Provider.id(this.config.name);
+    const tsData = await Config.Provider.get(tileSetId);
+
+    // initialize if not exist
+    if (tsData == null) this.import(tsData);
+
+    // Update if different
+    if (JSON.stringify(tsData) !== JSON.stringify(this.config)) this.import(tsData);
```
Comment (blacha) [inline src/provider.config.ts:82]: I don't think we can use stringy, there are timestamps and id's that may vary?

File: src/base.config.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      // initialize if not exist
+      this.import(newData);
+    } else if (!this.showDiff(newData, oldData)) {
+      // Update if different
+      this.import(newData);
+    }
+  }
+
+  showDiff(oldData: T, newData: T): boolean {
+    const ignoredProperties = ['id', 'createdAt', 'updatedAt'];
```
Comment (blacha) [inline src/base.config.ts:47]: should store this outside the function only declare it once

File: src/base.config.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    const changes = diff(oldData, newData, function (_path: string[], key: string) {
```
Comment (blacha) [inline src/base.config.ts:48]: could arrow function it

File: package.json [hunk 4]
```diff
@@ -6,10 +6,14 @@
...
   "license": "MIT",
   "dependencies": {
-    "@basemaps/cli": "^4.17.0",
-    "@basemaps/shared": "^4.17.0",
+    "@basemaps/geo": "^4.24.0",
+    "@basemaps/shared": "^4.24.0",
+    "@oclif/command": "^1.8.0",
     "@rushstack/ts-command-line": "4.7.6",
-    "pretty-json-log": "^0.3.2"
+    "@types/deep-diff": "^1.0.0",
```
Comment (blacha) [inline package.json:13]: `@types` should always be a dev dep

File: src/base.config.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
+import { LogType } from '@basemaps/shared';
+import { diff, Diff } from 'deep-diff';
+import * as c from 'ansi-colors';
+
+export const ignoredProperties = ['id', 'createdAt', 'updatedAt'];
```
Comment (blacha) [inline src/base.config.ts:5]: `TitleCase` constant exports

File: src/cli.import.config.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    tag: flags.string({ char: 't', description: 'PR tag(PR-number)', required: true }),
+    commit: flags.boolean({ description: 'Actually run job' }),
+  };
+
+  async run(): Promise<void> {
+    const logger = LogConfig.get();
+    const { flags } = this.parse(CommandImport);
+
+    Promise.all;
+    await importTileSet(flags.tag, flags.commit, logger);
```
Comment (blacha) [inline src/cli.import.config.ts:25]: TileSets should be the last thing imported.

File: src/provider.config.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+});
+
+export type ProviderConfigSchema = z.infer<typeof zProviderConfig>;
+
+export class ProviderUpdater extends Updater<ProviderConfigSchema, ConfigProvider> {
+  /**
+   * Class to apply an Provider source to the tile metadata db
+   * @param config a string or Provider to use
+   */
+  constructor(filename: string, config: unknown, tag: string, isCommit: boolean, logger: LogType) {
```
Comment (blacha) [inline src/provider.config.ts:48]: you can skip the consturctor if all it does is call super

File: src/imagery.config.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  async loadOldData(): Promise<ConfigImagery | null> {
+    const oldData = await Config.Imagery.get(this.config.id);
+    return oldData;
+  }
+
+  prepareNewData(oldData: ConfigImagery | null): ConfigImagery {
+    const now = Date.now();
+
+    // Tagging the id.
+    let id = `${this.config.id}@${this.tag}`;
```
Comment (blacha) [inline src/imagery.config.ts:54]: does this include the `im_` prefix? is there a `Config.Imagery.id` we could use?

File: src/tileset.config.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+  }
+  return {
+    r: parseHex(str.substr(0, 2)),
+    g: parseHex(str.substr(2, 2)),
+    b: parseHex(str.substr(4, 2)),
+    alpha: parseHex(str.substr(6, 2)),
+  };
+}
+
+const validateColor = (val: string): boolean => {
```
Comment (blacha) [inline src/tileset.config.ts:42]: use a `function` to be consistent with the other functions here.

File: src/cli.import.config.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+import { importProvider } from './provider.config';
+import { importStyle } from './style.conifg';
+import { importTileSet } from './tileset.config';
+
+export class CommandImport extends Command {
+  static description = 'Import Basemaps configs';
+
+  static flags = {
+    // add --version flag to show CLI version
+    version: flags.version({ char: 'v' }),
```
Comment (blacha) [inline src/cli.import.config.ts:13]: I thought these were defaults and don't need to be declared?

File: src/imagery.config.ts (hunk 8 above)
Comment (Wentao-Kuang) [reply src/imagery.config.ts:54]: @blacha , The id in the config file ${this.config.id} is already prefixed.

Comment (blacha) [reply src/imagery.config.ts:51]: Maybe we should take the `im_` prefix out of the config then, and use the `Config.Imagery.Id` Would be good to be consistent across the types.

Comment (blacha) [reply src/imagery.config.ts]: we removed the im_

**Review by blacha — APPROVED** [review]

**Pull request merged**